api/buildcraft/api/recipes/AssemblyRecipe.java

```diff
@@ -29,15 +32,60 @@ public AssemblyRecipe(NBTTagCompound nbt) {
         output = ItemStack.loadItemStackFromNBT(nbt.getCompoundTag("output"));
     }
 
+    public AssemblyRecipe(PacketBuffer buffer) throws IOException {
+        requiredMicroJoules = buffer.readLong();
+        ItemStack[] requiredStacksArray = new ItemStack[buffer.readInt()];
+        for(int i = 0; i < requiredStacksArray.length; i++) {
+            requiredStacksArray[i] = buffer.readItemStackFromBuffer();
+        }
+        requiredStacks = ImmutableSet.copyOf(requiredStacksArray);
+        output = buffer.readItemStackFromBuffer();
+    }
+
     public NBTTagCompound writeToNBT() {
         NBTTagCompound nbt = new NBTTagCompound();
         nbt.setLong("required_micro_joules", requiredMicroJoules);
         NBTTagList requiredStacksTag = new NBTTagList();
         for(ItemStack requiredStack : requiredStacks) {
-            requiredStacksTag.appendTag(requiredStack.writeToNBT(new NBTTagCompound()));
+            requiredStacksTag.appendTag(requiredStack.serializeNBT());
         }
         nbt.setTag("required_stacks", requiredStacksTag);
-        nbt.setTag("output", output.writeToNBT(new NBTTagCompound()));
+        nbt.setTag("output", output.serializeNBT());
         return nbt;
     }
+
+    public void writeToBuffer(PacketBuffer buffer) {
+        buffer.writeLong(requiredMicroJoules);
+        buffer.writeInt(requiredStacks.size());
+        requiredStacks.forEach(buffer::writeItemStackToBuffer);
+        buffer.writeItemStackToBuffer(output);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if(this == o) {
+            return true;
+        }
+        if(o == null || getClass() != o.getClass()) {
+            return false;
+        }
+
+        AssemblyRecipe that = (AssemblyRecipe) o;
+
+        if(requiredMicroJoules != that.requiredMicroJoules) {
+            return false;
+        }
+        if(requiredStacks != null ? !requiredStacks.equals(that.requiredStacks) : that.requiredStacks != null) {
```

Comment:
Unfortunately ItemStack doesn't implement "equals" so you have to expand this yourself. (Although you can use ````ItemStack.areItemStacksEqual```` to compare properly)